README.md
# @storyboard-fm/wavecore
A data structure for real-time reading, editing, seeking, and encoding of mono
WAV audio files via hypercore v10. Version control, branches, and peering come
for free thanks to the [`hypercore@next`][h] branch.
## Getting Started
### Installation
```sh
$ npm install git+ssh://git@github.com:storyboard-fm/wavecore.git
```
### Status
> WIP, expect breaking changes
## Background
This library's intent is to enable real-time peer-to-peer recording, editing, and
remixing of audio content without sacrificing fidelity, privacy, nor speed.

The WAV audio can be sourced from any valid instance of
[`random-access-storage`][ras] and can be therefore stored in memory as a
buffer, on S3 as a remote cloud URI, or as a file on the local file system. This
means it is functional on local offline-first client applications, server-side
applications, and web apps all from one codebase.
### Functionality
> ***ANYTHING YOU CAN DO IN HYPERCORE, YOU CAN DO IN WAVECORE.***
> [Start here!][h]
#### Create Operations
> A Wavecore can be created from a number of inputs and sources; with no
> arguments provided at all, a new Wavecore with sane defaults is provided.
- [x] Create new Wavecore with no inputs
- [x] Create new Wavecore with [`Source`][lmbsrc] input
- [x] Create new Wavecore with [hypercore][h] input
- [x] Create new Wavecore from other Wavecore
- [x] Create new Wavecore with [`random-access-storage`][ras] input
#### Write Operations
- [x] Append
- [x] Pad tail (Add blank data to tail to extend duration)
#### Editing Operations
> Wavecore audio editing tasks mimic the methods offered by [JavaScript
> Arrays][mdnarray]
- [x] Trim: [shift][shift]
- [x] Trim: [truncate][trunc]
- [x] [Split][split]
- [x] [Join][concat]
#### Timeline Operations
- [x] [Playback][play]
- [x] [Seeking][seek]
- [x] [Zero-cross detection][zero]
#### DSP Operations
- [x] [Gain][gain]
- [x] [Normalization][norm]
- [x] [Time Stretch][tempo]
#### Helper Functions
- [x] [Sessions][session]
- [x] [Snapshots][snapshot]
#### Analysis Operations
- [x] SoX `stat`
- [x] SoX `stats`
- [ ] SoX `spectrogram`
- [ ] `audiowaveform` peaks data
- [ ] `aubioonset` onset timing
- [ ] `fpcalc` fingerprint
#### Tagging Operations
- [x] RIFF [tags][tag]
- [ ] BWF tags
- [ ] Cue points (in progress)
#### P2P
- [ ] Replication between Wavecores
### Design
#### Mono WAV Only
There were several factors influencing the decision to support only mono WAV
data. Most notably, the Web Audio API's `AudioBuffer.getChannelData()` method,
as its name suggests, loads audio data one channel at a time. Wavecore audio
data can therefore be easily loaded inline with that method, querying it for
specific data ranges before having to allocate buffer resources.

By forcing all Wavecores to be mono-first, we also enable different processing to
occur on each channel of audio without worrying about data interleaving.
## Examples
### Recording Into A Wavecore
#### [`MediaRecorder`][mr]
##### Stream-based
```js
const mr = new MediaRecorder(stream)
const wave = new Wavecore()
wave.recStream(mr.stream)
```
##### Event-based
```js
const core = new Hypercore(ram)
mediaRecorder.onstop = function() {
  const wave = new Wavecore({ core })
}
mediaRecorder.ondataavailable = function(d) {
  core.append(d.data)
}
```
### Playing A Wavecore
#### Play Indexed Audio Data (SoX)
```js
const wavecore = new WavecoreSox({ source })
await wavecore.open()
wavecore.play({start: 5, end: 13}) // Play indeces 5 through 13
```
#### Listen For Live Audio Updates
> Listen to the audio being inserted in the Wavecore **in real-time!**
> ***Yes, you read that correctly: REAL-TIME!***
```js
const wavecore = new Wavecore({ source })
await wavecore.open()
const waveStream = wavecore._liveStream()
// Consume this ReadableStream to listen to the audio as it gets recorded into
// the Wavecore!
```
### Editing Operations
#### Editing Sessions & Snapshots
Here we make a snapshot of our Wavecore so we can test out an edit to the WAV
audio. By using our new session to test the edit we leave the original Wavecore
audio in-tact, enabling non-destructive editing with no additional memory
allocation necessary!
```js
const Wavecore = require('@storyboard-fm/wavecore')

const wave = new Wavecore({ source })
await wave.open()

const snapshot = wave.snapshot()
await wave.truncate(10)

console.log(wave.core.length) // 10
console.log(snapshot.core.length) // 58
```
#### Trimming Audio
##### Trim From Beginning
The following trims a Wavecore to start on the 20th index.
```js
const Wavecore = require('@storyboard-fm/wavecore')
const wave = new Wavecore({ source })

await wave.open()
console.log(wave.core.length) // 58
const shiftedCore = await Promise.resolve(wave.shift(20))
console.log(shiftedCore.length) // 38
```
##### Trim From End
The following truncates a Wavecore to the first 20 indeces.
```js
const Wavecore = require('@storyboard-fm/wavecore')
const wave = new Wavecore({ source })

await wave.open()
console.log(wave.core.length) // 58
await wave.truncate(20)
console.log(wave.core.length) // 20
```
#### Splitting Audio

Execute this script at `example2.js`.

```js
const Wavecore = require('.')

const w = new Wavecore({ source })

async function main() {
  await w.open()
  console.log('splitting Wavecore at index 22....')
  const [head, tail] = await w.split(22)
  console.log('done!', head, tail)
}

main()
```
### Signal Processing
#### Gain (+ Limiting) via SoX
> Increase gain by +6dBFS and add a limiter to prevent clipping
```js
const wave = new WavecoreSox({ source })
await wave.open()
const gainUp = await wave.gain('+6', {limiter:true})
```
#### Normalization
##### `AudioBuffer`
```js
const wave = new Wavecore({ source })
await wave.open()
const audioBufferNorm = await wave.audioBuffer({ normalize: true })
```
##### SoX
> Increase gain so that the peak dBFS value = 0
```js
const wave = new WavecoreSox({ source })
await wave.open()
const normCore = await wave.norm()
```
#### Playback Rate (SoX)
> Change the "tempo" of the audio without changing the pitch
```js
const wave = new WavecoreSox({ source })
await wave.open()
const snap = wave.snapshot() // save copy of original audio
const slowerWave = await Promise.resolve(wave.tempo(0.8)) // 20% slower
const fasterWave = await Promise.resolve(wave.tempo(1.1)) // 10% faster
```
## Tests

| Statements                  | Branches                | Functions                 | Lines             |
| --------------------------- | ----------------------- | ------------------------- | ----------------- |
| ![Statements](https://img.shields.io/badge/statements-80.11%25-yellow.svg?style=flat) | ![Branches](https://img.shields.io/badge/branches-76.55%25-red.svg?style=flat) | ![Functions](https://img.shields.io/badge/functions-79.76%25-red.svg?style=flat) | ![Lines](https://img.shields.io/badge/lines-85.61%25-yellow.svg?style=flat) |

We use `mocha`, with `nyc` for test coverage reporting.
```sh
$ npm run test
```

[concat]: https://storyboard-fm.github.io/wavecore/Wavecore.html#concat
[gain]: https://storyboard-fm.github.io/wavecore/WavecoreSox.html#gain
[h]: https://github.com/hypercore-protocol/hypercore-next
[lmbsrc]: https://storyboard-fm.github.io/little-media-box/Source.html
[mdnarray]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array
[mr]: https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder
[norm]: https://storyboard-fm.github.io/wavecore/WavecoreSox.html#norm
[play]:  https://storyboard-fm.github.io/wavecore/WavecoreSox.html#play
[ras]: https://github.com/random-access-storage
[seek]: https://storyboard-fm.github.io/wavecore/Wavecore.html#seek
[session]: https://storyboard-fm.github.io/wavecore/Wavecore.html#session
[snapshot]: https://storyboard-fm.github.io/wavecore/Wavecore.html#snapshot
[shift]: https://storyboard-fm.github.io/wavecore/Wavecore.html#shift
[sox]: http://sox.sourceforge.net/
[split]: https://storyboard-fm.github.io/wavecore/Wavecore.html#split
[tag]: https://storyboard-fm.github.io/wavecore/Wavecore.html#tag
[tempo]: https://storyboard-fm.github.io/wavecore/WavecoreSox.html#tempo
[trunc]: https://storyboard-fm.github.io/wavecore/Wavecore.html#truncate
[zero]: https://storyboard-fm.github.io/wavecore/Wavecore.html#_nextZero


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.3.1] - 2022-05-03
### Changed
- hypercore10 to latest version

## [0.3.0] - 2022-04-12
### Added
- more opts for audiobuffer to help conform to audio source properties

## [0.2.9] - 2022-04-01
### Changed
- hypercore10 to latest version
- hypercore10 to latest version

## [0.2.7] - 2022-03-22
### Changed
- class index sizes to match exact time values for WAV format

## [0.2.6] - 2022-03-22
### Changed
- hypercore10 to latest version

## [0.2.5] - 2022-03-16
### Fixed
- the Wavecore class namespace bundle to be up to date with latest source

## [0.2.4] - 2022-03-16
### Changed
- release version because i fuxed up the release process

## [0.2.3] - 2022-03-16
### Changed
- Wavecore class to extend Hypercore directly

### Fixed
- all tests and examples that were previously failing due to the class extension change

## [0.2.2] - 2022-03-15
### Added
- web audio api node dev package

## [0.2.1] - 2022-03-14
### Changed
- hypercore10 to latest version with stabilised replicator code

## [0.2.0] - 2022-03-09
### Added
- start and end opts for audioBuffer

### Fixed
- failing unit tests due to sox lib

### Removed
- WaveFile dependency

## [0.1.9] - 2022-03-09
### Changed
- class structure so that sox commands are in a class extension, WavecoreSox

## [0.1.8] - 2022-03-08
### Changed
- stream recording to use same indeces as raw buffers

## [0.1.7] - 2022-03-08
### Changed
- open() method to read from the source at specified intervals; removes dependency on lmb

## [0.1.6] - 2022-03-04
### Added
- audiobuffer mixing
- classification method to determine whether audio is quiet or a voice recording

## [0.1.5] - 2022-03-03
### Added
- encryptionKey opt to wavecore constructor

## [0.1.4] - 2022-03-01
### Added
- normalize and dcOffset DSP options for audiobuffer method

### Fixed
- incorrect formatting opts made audiobuffer output wrong

## [0.1.3] - 2022-02-19
### Added
- seek optional argument to return next zero crossing byteOffset

## [0.1.2] - 2022-02-17
### Changed
- nextZero method to accept a seek return value

## [0.1.1] - 2022-02-17
### Added
- vad method for removing excessive silences

## [0.1.0] - 2022-02-16
### Changed
- volume adjustment and normalization methods to allow for a custom range of indeces to process, rather than the entire wavecore

### Added
- new methods for creating a wavecore from a new recording via the `rec` command line tool

## [0.0.9] - 2022-02-16
### Added
- Gain method to apply or attenuate gain in the audio

## [0.0.8] - 2022-02-15
### Fixed
- open method so it does not open cores twice

### Added
- a class method to record an audio stream into a Wavecore

## [0.0.7] - 2022-02-15
### Changed
- some functions to class getters, improving performance

## [0.0.6] - 2022-02-15
### Added
- custom playback range option for playback method

## [0.0.5] - 2022-02-14
### Added
- Wavecore class and associated little-media-box dependencies
- git hooks for linting before commit
- Seek the wave file via hypercore
- shift() method for trimming wavecores from the beginning
- github actions workflows for automated testing
- new jsdoc template
- addBlank() method for adding blank data to end of file
- audiobuffer and wav file support
- Tempo method for time stretching audio
- has method to check if a wavecore has a block at a given index
- stats option for tempo method
- storage option for wav method

### Changed
- the return of the snapshot method from a hypercore to a wavecore

[Unreleased]: https://github.com/Storyboard-fm/wavecore/compare/v0.3.1...HEAD
[0.3.1]: https://github.com/Storyboard-fm/wavecore/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.9...v0.3.0
[0.2.9]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.7...v0.2.9
[0.2.7]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.6...v0.2.7
[0.2.6]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.5...v0.2.6
[0.2.5]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.4...v0.2.5
[0.2.4]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.3...v0.2.4
[0.2.3]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/Storyboard-fm/wavecore/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.9...v0.2.0
[0.1.9]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.8...v0.1.9
[0.1.8]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.7...v0.1.8
[0.1.7]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.6...v0.1.7
[0.1.6]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.5...v0.1.6
[0.1.5]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.4...v0.1.5
[0.1.4]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/Storyboard-fm/wavecore/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/Storyboard-fm/wavecore/compare/v0.0.9...v0.1.0
[0.0.9]: https://github.com/Storyboard-fm/wavecore/compare/v0.0.8...v0.0.9
[0.0.8]: https://github.com/Storyboard-fm/wavecore/compare/v0.0.7...v0.0.8
[0.0.7]: https://github.com/Storyboard-fm/wavecore/compare/v0.0.6...v0.0.7
[0.0.6]: https://github.com/Storyboard-fm/wavecore/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/Storyboard-fm/wavecore/releases/tags/v0.0.5


sox.js
const Hypercore = require('hypercore')
const nanoprocess = require('nanoprocess')
const { PassThrough } = require('stream')
const process = require('process')
const ram = require('random-access-memory')
const Wavecore = require('./index')
const WaveFile = require('wavefile').WaveFile

/**
 * The `WavecoreSox` class provides a SoX interface for Wavecore.
 * @class
 * @extends {Wavecore}
 */
class WavecoreSox extends Wavecore {
  /**
   * The `WavecoreSox` class constructor.
   * @arg {Object} [opts={}] - Options
   * @arg {Hypercore} [opts.core=null] - Provide a previously-made hypercore.
   * @arg {Integer} [opts.indexSize=null] - Declare an alternate index size.
   * @arg {Buffer|Readable|PassThrough|Array|ArrayBuffer} [opts.source] - PCM
   * source.
   * @arg {Buffer} [opts.encryptionKey=null] - Encryption key
   * @arg {random-access-storage} [opts.storage=ram] - Provide storage instance.
   * @returns {WavecoreSox} wavecoreSox - The new Wavecore with SoX methods.
   */
  constructor(opts={
    core: null,
    key: null,
    encryptionKey: null,
    hypercoreOpts: null,
    indexSize: 96000,
    parent: null,
    source: null,
    storage: null
  }) {
    super(opts)
  }
  /**
   * Get the maximum volume adjustment value for the Wavecore's PCM audio data.
   * Used by the `norm()` method to ensure the normalized audio does not clip.
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Number} [opts.start=0] - Index from which to start the stream
   * @arg {Number} [opts.end=-1] - Index where the stream should end.
   * @returns {Number} vol - The SoX `vol -v` value.
   */
  _volAdjust(opts = { start: 0, end: -1 }) {
    const { start, end } = opts
    return new Promise((resolve, reject) => {
      const statsCmd = nanoprocess('sox', [
        '-r',
        '48000',
        '-b',
        '16',
        '-e',
        'signed',
        '-t',
        'raw',
        '-',
        '-n',
        'stat',
        '-v',
      ])
      statsCmd.open((err) => {
        if (err) reject(err)

        const statsOut = []

        const pt = new PassThrough()
        pt.on('data', (d) => statsOut.push(`${d}`))

        statsCmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-zero exit code'))
          resolve(Number(statsOut.join('')))
        })
        statsCmd.stderr.pipe(pt)

        this._rawStream(start, end).pipe(statsCmd.stdin)
      })
    })
  }
  /**
   * Apply gain attenuation or amplification to the Wavecore audio.
   * @arg {String} g - A string beginning with `+` or `-` indicating gain
   * operation to perform on the audio, i.e., `+8` or `-2.3`.
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Number} [opts.start=0] - Start index
   * @arg {Number} [opts.end=-1] - End index
   * @arg {Boolean} [opts.limiter=false] - Whether to apply a limiter to the
   * gain function to prevent clipping.
   * @returns {Wavecore} - New Wavecore with the gain processing applied.
   */
  async gain(g, opts = { start: 0, end: -1, limiter: false }) {
    const { start, end, limiter } = opts
    const rs = this._rawStream(start, end)
    const cmdOpts = [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
      '-t',
      'raw',
      '-',
      'gain',
    ]
    if (limiter) cmdOpts.push('-l')
    cmdOpts.push(`${g}`)

    const gainCmd = nanoprocess('sox', cmdOpts)

    const prom = new Promise((resolve, reject) => {
      gainCmd.open((err) => {
        if (err) reject(err)

        const newGainCore = new WavecoreSox(ram)
        const ws = newGainCore.createWriteStream({
          highWaterMark: this.indexSize,
        })
        ws.on('close', () => {
          newGainCore
            .update()
            .then(() => resolve(newGainCore))
        })
        gainCmd.stdout.pipe(ws)
        rs.pipe(gainCmd.stdin)
      })
    })

    return await Promise.resolve(prom)
  }
  /**
   * Listen live to the audio data coming in to the Wavecore. Great way to
   * monitor the audio inputs or broadcast the content to others.
   */
  monitor() {
    const cmdOpts = [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
    ]
    const playCmd = nanoprocess('play', cmdOpts)
    playCmd.open((err) => {
      playCmd.stdout.pipe(process.stdout)
      this.liveStream.pipe(playCmd.stdin)
    })
  }
  /**
   * Normalize the audio data in the Wavecore. Returns a new Wavecore instance.
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Number} [opts.start=0] - Index from which to start the stream
   * @arg {Number} [opts.end=-1] - Index where the stream should end.
   */
  async norm(opts = { start: 0, end: -1 }) {
    const { start, end } = opts
    try {
      const vol = await this._volAdjust({ start, end })
      const normCmd = nanoprocess('sox', [
        '-r',
        '48000',
        '-b',
        '16',
        '-e',
        'signed',
        '-t',
        'raw',
        '-',
        '-t',
        'raw',
        '-',
        'vol',
        vol,
      ])

      const rs = this._rawStream(start, end)

      const prom = new Promise((resolve, reject) => {
        normCmd.open((err) => {
          if (err) reject(err)

          // TODO figure out why number of indeces higher in new wavecore
          const newCore = new WavecoreSox(ram)

          const pt = new PassThrough({ highWaterMark: this.indexSize })
          pt.on('error', (err) => reject(err))
          pt.on('data', (d) => newCore.append(d))

          normCmd.on('close', (code) => {
            if (code !== 0) reject(new Error('Non-zero exit code'))
            resolve(newCore)
          })
          normCmd.stdout.pipe(pt)

          rs.pipe(normCmd.stdin)
        })
      })
      return await Promise.resolve(prom)
    } catch (err) {
      throw err
    }
  }
  /**
   * Play the raw Wavecore PCM audio via a nanoprocess
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Number} [opts.start=0] - Start index
   * @arg {Number} [opts.end=-1] - End index
   * @arg {nanoprocess} [opts.np=null] - Declare a custom nanoprocess for playback
   * @see {@link https://github.com/little-core-labs/nanoprocess nanoprocess}
   */
  play(opts = { start: 0, end: -1, np: null }) {
    const { np, start, end } = opts

    let proc = null

    if (np) {
      proc = np
    } else {
      proc = nanoprocess('play', [
        '-r',
        '48000',
        '-b',
        '16',
        '-e',
        'signed',
        '-t',
        'raw',
        '-',
      ])
    }

    if (!proc) throw new Error('nanoprocess didnt work wtf')
    const rs = this._rawStream(start, end)

    proc.open((err) => {
      if (err) throw err
      rs.on('end', () => console.log('ended'))
      proc.stderr.pipe(process.stderr)
      proc.stdout.pipe(process.stdout)
      rs.pipe(proc.stdin)
    })
  }
  /** Record into the Wavecore via the `rec` CLI application.
   * @arg {String} [dur="30:00"] - Duration string for recording; defaults to
   * 30min.
   */
  async _rec(dur = '30:00') {
    const cmdOpts = [
      '-r',
      '48000',
      '-c',
      '1',
      '-b',
      '16',
      '-e',
      'signed-integer',
      '-t',
      'raw',
      '-',
      'trim',
      '0',
      `${dur}`,
    ]
    const recCmd = nanoprocess('rec', cmdOpts)
    const prom = new Promise((resolve, reject) => {
      recCmd.open((err) => {
        if (err) reject(err)

        recCmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-Zero exit code!', code))

          this.update().then(() => resolve())
        })

        recCmd.stdout.pipe(
          this.createWriteStream({ highWaterMark: this.indexSize })
        )
      })
    })

    await Promise.resolve(prom)
  }
  /**
   * Runs `sox -n stats` on the raw audio in the Wavecore, via a nanoprocess.
   * @async
   * @arg {Object} [opts={}] - Optional opts object for declaring index
   * @arg {Number} [opts.index=null] - Declare index to get stats on
   * @returns {String} statsOut - The string of stats information returned by
   * SoX
   */
  async stats(opts = { index: null }) {
    const { index } = opts
    const statsCmd = nanoprocess('sox', [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
      '-n',
      'stats',
      'stat',
    ])

    const statsOut = []

    const pt = new PassThrough()
    pt.on('data', (d) => statsOut.push(`${d}`))
    const prom = new Promise((resolve, reject) => {
      statsCmd.open((err) => {
        if (err) reject(err)

        statsCmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-zero exit code'))
          resolve(statsOut.join(''))
        })
        statsCmd.stderr.pipe(pt)

        let rs = null

        if (index !== null) {
          rs = this._rawStream(index, index + 1)
        } else {
          rs = this.createReadStream()
        }

        rs.pipe(statsCmd.stdin)
      })
    })
    return await Promise.resolve(prom)
  }
  /**
   * Increase or decrease playback speed of audio samples, without changing the
   * pitch of the audio itself. Returns a new Wavecore containing the
   * time-stretched samples.
   * @arg {Float} f - The new tempo factor. 0.9 = slow down by 10%; 1.1 = faster
   * by 10%.
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Boolean} [opts.stats=false] - Whether to also get SoX stats on the
   * time-stretched audio data. Currently these stats are output to `stdout`.
   * Useful for getting the new Wavecore's audio duration.
   * @returns {Wavecore} stretchedCore - The new time-stretched Wavecore.
   */
  async tempo(f, opts = { stats: false }) {
    const { stats } = opts
    const cmdOpts = [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
      '-t',
      'raw',
      '-',
      'tempo',
      '-s',
      `${f}`,
    ]
    if (stats) cmdOpts.push('stats')
    const tempoCmd = nanoprocess('sox', cmdOpts)
    const prom = new Promise((resolve, reject) => {
      tempoCmd.open((err) => {
        if (err) reject(err)

        const newCore = new WavecoreSox(ram)

        const pt = new PassThrough()
        pt.on('data', (d) => newCore.append(d))

        tempoCmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-zero exit code'))
          resolve(newCore)
        })
        tempoCmd.stdout.pipe(pt)
        if (stats) tempoCmd.stderr.pipe(process.stdout)

        let rs = this._rawStream()
        rs.pipe(tempoCmd.stdin)
      })
    })
    return await Promise.resolve(prom)
  }
  /**
   * Runs `sox vad` on the Wavecore audio. Trims excessive silence from the
   * front of voice recordings.
   * @returns {Wavecore}
   */
  async vad() {
    const cmdOpts = [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
      '-t',
      'raw',
      '-',
      'vad',
    ]
    const cmd = nanoprocess('sox', cmdOpts)
    const newCore = new WavecoreSox(ram)
    const normCore = await this.norm()
    const prom = new Promise((resolve, reject) => {
      cmd.open((err) => {
        if (err) reject(err)

        cmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-zero exit!', code))
          resolve(newCore)
        })

        cmd.stdout.pipe(
          newCore.createWriteStream({ highWaterMark: this.indexSize })
        )
        normCore._rawStream().pipe(cmd.stdin)
      })
    })
    return await Promise.resolve(prom)
  }
  /**
   * Returns a `Promise` which resolves a `Buffer` of a PCM WAV file. Requires
   * `sox` in PATH.
   * @arg {Object} [opts={}] - Optional options object
   * @arg {Boolean} [opts.store=false] - Whether to store the wav as a buffer in
   * the Wavecore class instance.
   * @returns {Buffer} wavBuf - WAV file Buffer
   */
  async wav(opts = { store: false }) {
    const { store } = opts
    const bufs = []
    const pt = new PassThrough()
    pt.on('error', (err) => reject(err))
    pt.on('data', (d) => bufs.push(d))
    const soxCmd = nanoprocess('sox', [
      '-r',
      '48000',
      '-b',
      '16',
      '-e',
      'signed',
      '-t',
      'raw',
      '-',
      '-t',
      'wav',
      '-',
    ])
    const np = new Promise((resolve, reject) => {
      soxCmd.open((err) => {
        if (err) reject(err)

        soxCmd.on('close', (code) => {
          if (code !== 0) reject(new Error('Non-Zero exit-code!', code))
          const wavBuf = Buffer.concat(bufs)
          if (store) this.wavBuffer = wavBuf
          if (this.wavBuffer) this.wavFile = new WaveFile(this.wavBuffer)
          if (this.wavFile) {
            Array.from(this.tags).forEach((t) =>
              this.wavFile.setTag(t[0], t[1])
            )
            this.wavBuffer = Buffer.from(this.wavFile.toBuffer())
          }
          resolve(wavBuf)
        })
        soxCmd.stdout.pipe(pt)
        const rs = this.createReadStream()
        rs.pipe(soxCmd.stdin)
      })
    })
    return await Promise.resolve(np)
  }
}

module.exports = WavecoreSox
